### PR TITLE
Check duplicate constraints with tuple name differences

### PR DESCRIPTION
**Pull request by jcouv**
The current check for duplicates in constraints does a full type comparison, so no error is reported.
This PR checks for duplicate constraints ignoring tuple names.

```C#
interface I<T> { T Get(); }
class C<U> where U : I<(int a, int b)>, I<(int notA, int notB)> // this is currently allowed, and would become an error
{
    void M(U u)
    {
        var mystery = u.Get(); // this is currently (int notA, int notB)
    }
}
```

The problem does not exist in VB.
I started a discussion with compat council to review this back compat break (from 7.0).

@dotnet/roslyn-compiler for review.

Fixes https://github.com/dotnet/roslyn/issues/17994

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 1]
```diff
@@ -21749,6 +21749,71 @@ class Derived : Base<I2<(int notA, int notB)>> {}
...
+
+class C<U> where U : I1<dynamic> {}
+";
+            var comp = CreateCompilationWithMscorlib(source, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics(
+                // (4,22): error CS1968: Constraint cannot be a dynamic type 'I1<dynamic>'
+                // class C<U> where U : I1<dynamic> {}
+                Diagnostic(ErrorCode.ERR_ConstructedDynamicTypeAsBound, "I1<dynamic>").WithArguments("I1<dynamic>").WithLocation(4, 22)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: Is this test related to the issue? (There are existing tests for this error.)

Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: There is one other test (`CS1968ERR_ConstructedDynamicTypeAsBound`), but it covers more complex cases.

Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: The reason this is related to the issue is that the comparison for types can ignore tuple names, dynamic, modopts, or a combination of the above.
In my change, I only ignore differences in tuple names, since dynamic is already disallowed.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 2]
```diff
@@ -21749,6 +21749,71 @@ class Derived : Base<I2<(int notA, int notB)>> {}
...
+        {
+            var source = @"
+interface I1<T> {}
+interface I2<T> : I1<T> {}
+
+class C<U> where U : I1<(int a, int b)>, I2<(int notA, int notB)> {}
+";
+            var comp = CreateCompilationWithMscorlib(source, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: Why is the indirect case not an error?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: I see, thanks. But to hit the new code path, wouldn't we need two constraints? `C<U> where U : I1<dynamic>, I1<object>`

Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: Since dynamic is blocked before the comparison I modified, the second constraint doesn't change the result or hit the new path (dynamic gets caught first).

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 3]
```diff
@@ -1471,5 +1471,6 @@ internal enum ErrorCode
         ERR_InvalidPreprocessingSymbol = 8301,
         ERR_FeatureNotAvailableInVersion7_1 = 8302,
         ERR_LanguageVersionCannotHaveLeadingZeroes = 8303,
+        ERR_DuplicateBoundWithDifferentTupleNames = 8304,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: >ERR_DuplicateBoundWithDifferentTupleNames = 8304 [](start = 8, length = 48)

Is this going to conflict with an error for a new feature? Perhaps we can use the gap between 8209 and 8300 for bug fixes?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 4]
```diff
@@ -21749,6 +21749,71 @@ class Derived : Base<I2<(int notA, int notB)>> {}
...
+        public void TupleNameDifferencesInIndirectDuplicateConstraint()
+        {
+            var source = @"
+interface I1<T> {}
+interface I2<T> : I1<T> {}
+
+class C<U> where U : I1<(int a, int b)>, I2<(int notA, int notB)> {}
+";
+            var comp = CreateCompilationWithMscorlib(source, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: >comp.VerifyDiagnostics(); [](start = 12, length = 25)

Yes, it feels like there should be an error for this case. We get one for ```class C : I1<(int a, int b)>, I2<(int notA, int notB)>{}```.

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 5]
```diff
@@ -19174,6 +19174,71 @@ End Class
...
+Interface I2(Of T)
+    Inherits I1(Of T)
+End Interface
+
+Class C(Of U As {I1(Of (a As Integer, b As Integer)), I2(Of (notA As Integer, notB As Integer))})
+End Class
+]]></file>
+</compilation>, additionalRefs:=s_valueTupleRefs)
+
+            compilation.AssertTheseDiagnostics()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: >compilation.AssertTheseDiagnostics() [](start = 12, length = 36)

It feels like this should be an error.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs (hunk 2 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: @cston @AlekseyTs @gafter I'm looking into this more and something bugs me.
A constraint definition (type with a list of implemented interfaced) is similar to a type definition, and it seems functionally very similar too (it tells you what interfaces to expect when you have an instance of that type).
But, as far as I can tell, the checks that we apply in both cases are different. In the constraint scenario, it looks like we don't do type unification checks. For instance, there is no error in `class C<U, V> where U : I<V>, I<int> { }`, but there is in `class C2<V> : I<V>, I<int> { }`.
Any ideas why?

Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: From discussion with Aleksey, I now understand why the checks would be different between those two scenarios. One declares a new types and so much enforce CLR rules, while the other will be using an existing type (necessarily already checked elsewhere) so it can be more relaxed.

**Comment by jcouv** [comment]
From discussion with LDM, we're lowering the priority of this fix. If we do it, we'd want to do the deep check too. I'll close the PR for now.
